lib/ClangImporter/ImportDecl.cpp

```diff
@@ -2940,10 +2940,17 @@ namespace {
       // Add the implicit 'self' parameter patterns.
       SmallVector<ParameterList *, 4> bodyParams;
       auto selfVar =
-        ParamDecl::createSelf(SourceLoc(), dc,
-                              /*isStatic*/
-                              decl->isClassMethod() || forceClassMethod);
+        ParamDecl::createSelf(SourceLoc(), dc, /*isStatic*/!isInstance);
       bodyParams.push_back(ParameterList::createWithoutLoc(selfVar));
+      Type selfInterfaceType;
+      if (dc->getAsProtocolOrProtocolExtensionContext()) {
```

Comment:
These four lines appear in many places -- it would be awesome if you factored it out